WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java

```diff
@@ -98,75 +106,105 @@ public int onStartCommand(Intent intent, int flags, int startId) {
-            completeCurrentUpload();
+            completeUploadWithId(event.media.getId());
             uploadNextInQueue();
         } else if (event.completed) {
             // Upload completed
             AppLog.i(AppLog.T.MEDIA, "Upload completed - localId=" + event.media.getId() + " title=" + event.media.getTitle());
-            mCurrentUpload.setMediaId(event.media.getMediaId());
-            completeCurrentUpload();
+            // here we need to edit the corresponding post
+            updatePostWithMediaUrl(event.media);
+            completeUploadWithId(event.media.getId());
             uploadNextInQueue();
+            stopServiceIfUploadsComplete();
         } else {
             // Upload Progress
-            // TODO check if we need to broadcast event.media, event.progress or we're just fine with
+            // TODO check if we need to re-broadcast event.media, event.progress or we're just fine with
             // listening to  event.media, event.progress
+            AppLog.d(AppLog.T.MEDIA, event.media.getId() + " - progressing " + event.progress);
         }
     }
 
     private void handleOnMediaUploadedError(@NonNull OnMediaUploaded event) {
         AppLog.w(AppLog.T.MEDIA, "Error uploading media: " + event.error.message);
         // TODO: Don't update the state here, it needs to be done in FluxC
-        mCurrentUpload.setUploadState(UploadState.FAILED.name());
-        mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(mCurrentUpload));
-        completeCurrentUpload();
-        // TODO: check whether we need to broadcast the error or maybe it is enough to register for FluxC events
-        // event.media, event.error
+        MediaModel media = getMediaFromQueueById(event.media.getId());
+        if (media != null) {
+            media.setUploadState(UploadState.FAILED.name());
+            mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(media));
+        }
+        completeUploadWithId(event.media.getId());
         uploadNextInQueue();
     }
 
-    private void uploadNextInQueue() {
-        // waiting for response to current upload request
-        if (mCurrentUpload != null) {
-            AppLog.i(AppLog.T.MEDIA, "Ignoring request to uploadNextInQueue, only one media item can be uploaded at a time.");
-            return;
+    private synchronized void updatePostWithMediaUrl(MediaModel media){
+        if (media != null) {
+            PostModel post = mPostStore.getPostByLocalPostId(media.getPostId());
+            if (post != null) {
+                String content = post.getContent();
+                if (content != null) {
+
+                    // actually replace the media ID with the media uri
+                    MediaUploadReadyProcessor processor = new MediaUploadReadyProcessor();
+                    PostModel modifiedPost = processor.replaceMediaFileWithUrlInPost(post, String.valueOf(media.getId()), FluxCUtils.mediaFileFromMediaModel(media));
```

Comment:
This seems unsafe - the post can still be open in the editor when this happens, and we can't guarantee that the editor will receive the `OnMediaUploaded` event before the media service does. In that case we' might have some data loss.

As things are currently implemented, we would need a way to tell if the post is currently open in the editor, and in that case leave it up to the editor to make the substitution.